pkg/bgpv1/gobgp/server.go

```diff
@@ -117,39 +117,58 @@ func (g *GoBGPServer) AddNeighbor(ctx context.Context, n types.NeighborRequest)
 
 // UpdateNeighbor will update the existing CiliumBGPNeighbor in the gobgp.BgpServer.
 func (g *GoBGPServer) UpdateNeighbor(ctx context.Context, n types.NeighborRequest) error {
-	peer, err := g.getPeerConfig(ctx, n.Neighbor, true)
+	peer, needsFullReset, err := g.getPeerConfig(ctx, n.Neighbor, true)
 	if err != nil {
 		return err
 	}
+
+	// update peer config
 	peerReq := &gobgp.UpdatePeerRequest{
-		DoSoftResetIn: true, // should perform soft reset only if needed
-		Peer:          peer,
+		Peer: peer,
 	}
-	if _, err = g.server.UpdatePeer(ctx, peerReq); err != nil {
+	updateRes, err := g.server.UpdatePeer(ctx, peerReq)
+	if err != nil {
 		return fmt.Errorf("failed while updating peer %v %v: %w", n.Neighbor.PeerAddress, n.Neighbor.PeerASN, err)
 	}
+
+	// perform full / soft peer reset if necessary
+	if needsFullReset || updateRes.NeedsSoftResetIn {
+		g.logger.Infof("Resetting peer %s (ASN %d) due to a config change", peer.Conf.NeighborAddress, peer.Conf.PeerAsn)
+		resetReq := &gobgp.ResetPeerRequest{
+			Address:       peer.Conf.NeighborAddress,
+			Communication: "Peer configuration changed",
+		}
+		if !needsFullReset {
+			resetReq.Soft = true
+			resetReq.Direction = gobgp.ResetPeerRequest_IN
+		}
+		if err = g.server.ResetPeer(ctx, resetReq); err != nil {
```

Comment:
Do we have any situation that needs a soft reset now? IIUC, a soft reset is needed when we update the route filter or something that needs a peer to resend the route. In our current use case, it will never happen, right?

Also, I'd like to know if

Reply:
Right, at the moment we do not have a case for for it, I put it here mostly to have it covered correctly for the future.